### PR TITLE
gitlab-ci: add lv-cuda

### PR DESCRIPTION
**Pull request by jedbrown**

### CODE REVIEW

**Comment by jeremylt** [comment]
Bug in `junit.py` - t318 should be skipped for both `/**/magma` and `/gpu/cuda/ref`

**Comment by jedbrown** [comment]
Cool, and then we should add PETSc, which can be an external install managed by you or me. Preference?

**Comment by jeremylt** [comment]
I can definitely make a PETSc with CUDA build on lv tomorrow

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -104,7 +104,7 @@ lv-cuda:
 # -- PETSc with CUDA (minimal)
     - export PETSC_DIR=/home/jeth8984/petsc PETSC_ARCH=cuda-O && git -C $PETSC_DIR describe
     - echo "-------------- PETSc ---------------" && make -C $PETSC_DIR info
-    - make -k -j$NPROC_GPU BACKENDS="$BACKENDS_GPU" junit search="petsc fluids solids"
```
Comment (jeremylt) [inline .gitlab-ci.yml]: Fluids and Solids are fairly resource heavy - I think we might not be able to test them here

Comment (jedbrown) [reply .gitlab-ci.yml]: What metric are you using? What would be the marginal cost to do one run of each in this environment?

Comment (jeremylt) [reply .gitlab-ci.yml]: Before I dropped these two I got a bunch of`CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES` error messages. It might be easier to adjust this over in Leila's branch where we aren't constantly pulling vectors back and forth between host and device.

File: examples/petsc/Makefile [hunk 2]
```diff
@@ -37,34 +37,35 @@ LDLIBS = $(call pkgconf, --libs-only-l $(PETSc.pc) $(ceed.pc)) -lm
...
 utils.c := $(sort $(wildcard $(SRCDIR)/*.c))
 utils.o = $(utils.c:%.c=$(OBJDIR)/%.o)
-
-all: area bps bpsraw bpssphere multigrid
+utils: $(utils.o)
 
 area.c := area.c
 area.o = $(area.c:%.c=$(OBJDIR)/%.o)
-area: $(area.o) $(utils.o) | $(PETSc.pc) $(ceed.pc)
+area: $(area.o) $(utils.o) | utils $(PETSc.pc) $(ceed.pc)
```
Comment (jedbrown) [inline examples/petsc/Makefile:48]: Why is this necessary? We already have `$(utils.o)` as a normal dependency so why add a phony order-only dependency?

Comment (jedbrown) [reply examples/petsc/Makefile:48]: Is there a way to reproduce it (even if not every run) outside of CI? If this works (big if), I think it's masking the real issue and encourages misunderstandings of how make works. I'd like to find a non-hacky solution, though it doesn't need to be in this PR.

Comment (jeremylt) [reply examples/petsc/Makefile:48]: The build failure occurs ~5% of the time when I run `make prove -j search=petsc` on my machine.

<details>

```
make -C examples/petsc CEED_DIR=`pwd` \
  PETSC_DIR="/home/jeremy/Dev/petsc" OPT="-O -g -march=native -ffp-contract=fast -fopenmp-simd" bpsraw
make -C examples/petsc CEED_DIR=`pwd` \
  PETSC_DIR="/home/jeremy/Dev/petsc" OPT="-O -g -march=native -ffp-contract=fast -fopenmp-simd" area
make -C examples/petsc CEED_DIR=`pwd` \
  PETSC_DIR="/home/jeremy/Dev/petsc" OPT="-O -g -march=native -ffp-contract=fast -fopenmp-simd" multigrid
make -C examples/petsc CEED_DIR=`pwd` \
  PETSC_DIR="/home/jeremy/Dev/petsc" OPT="-O -g -march=native -ffp-contract=fast -fopenmp-simd" bps
make -C examples/petsc CEED_DIR=`pwd` \
  PETSC_DIR="/home/jeremy/Dev/petsc" OPT="-O -g -march=native -ffp-contract=fast -fopenmp-simd" bpssphere
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/petsc'
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/petsc'
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/petsc'
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/petsc'
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/petsc'
          CC build/bpsraw.o
          CC build/bps.o
          CC build/multigrid.o
          CC build/area.o
          CC build/bpssphere.o
          CC build/src/bcfunctions.o
          CC build/src/bcfunctions.o
          CC build/src/bcfunctions.o
          CC build/src/bcfunctions.o
          CC build/src/bcfunctions.o
          CC build/src/libceedsetup.o
          CC build/src/libceedsetup.o
          CC build/src/libceedsetup.o
          CC build/src/libceedsetup.o
          CC build/src/libceedsetup.o
          CC build/src/matops.o
          CC build/src/matops.o
          CC build/src/matops.o
          CC build/src/matops.o
          CC build/src/matops.o
          CC build/src/petscutils.o
          CC build/src/petscutils.o
          CC build/src/petscutils.o
          CC build/src/petscutils.o
          CC build/src/petscutils.o
      LINK.o ./bpssphere
      LINK.o ./multigrid
      LINK.o ./bps
      LINK.o ./bpsraw
      LINK.o ./area
/usr/bin/ld: build/bpssphere.o: in function `main':
/home/jeremy/Dev/libCEED/examples/petsc/bpssphere.c:170: undefined reference to `SetupDMByDegree'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/bpssphere.c:234: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/bpssphere.c:243: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/bpssphere.c:164: undefined reference to `ProjectToUnitSphere'
/usr/bin/ld: build/src/libceedsetup.o: in function `SetupLibceedByDegree':
/home/jeremy/Dev/libCEED/examples/petsc/src/libceedsetup.c:71: undefined reference to `CreateRestrictionFromPlex'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/libceedsetup.c:74: undefined reference to `CreateRestrictionFromPlex'
/usr/bin/ld: build/src/matops.o: in function `MatGetDiag':
/home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:21: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:28: undefined reference to `MemTypeP2C'
/usr/bin/ld: build/src/matops.o: in function `ApplyLocal_Ceed':
/home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:57: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:58: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:64: undefined reference to `MemTypeP2C'
/usr/bin/ld: build/src/matops.o:/home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:65: more undefined references to `MemTypeP2C' follow
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:63: bpssphere] Error 1
rm build/src/.DIR build/.DIR
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/petsc'
make: *** [Makefile:525: build/petsc-bpssphere] Error 2
make: *** Waiting for unfinished jobs....
/usr/bin/ld: build/multigrid.o: in function `main':
/home/jeremy/Dev/libCEED/examples/petsc/multigrid.c:180: undefined reference to `Kershaw'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/multigrid.c:239: undefined reference to `SetupDMByDegree'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/multigrid.c:312: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/multigrid.c:334: undefined reference to `MemTypeP2C'
/usr/bin/ld: build/src/libceedsetup.o: in function `SetupLibceedByDegree':
/home/jeremy/Dev/libCEED/examples/petsc/src/libceedsetup.c:71: undefined reference to `CreateRestrictionFromPlex'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/libceedsetup.c:74: undefined reference to `CreateRestrictionFromPlex'
/usr/bin/ld: build/src/matops.o: in function `MatGetDiag':
/home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:21: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:28: undefined reference to `MemTypeP2C'
/usr/bin/ld: build/src/matops.o: in function `ApplyLocal_Ceed':
/home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:57: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:58: undefined reference to `MemTypeP2C'
/usr/bin/ld: /home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:64: undefined reference to `MemTypeP2C'
/usr/bin/ld: build/src/matops.o:/home/jeremy/Dev/libCEED/examples/petsc/src/matops.c:65: more undefined references to `MemTypeP2C' follow
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:68: multigrid] Error 1
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/petsc'
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/petsc'
make: *** [Makefile:525: build/petsc-multigrid] Error 2
cp examples/petsc/bpsraw build/petsc-bpsraw
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/petsc'
cp examples/petsc/bps build/petsc-bps
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/petsc'
cp examples/petsc/area build/petsc-area
```
<\details>

Comment (jedbrown) [reply examples/petsc/Makefile:48]: Oh yeah, recursive make is literally the devil. Each submake is building the utils independently (overwriting each others results). Possible solutions:
1. make an `examples/petsc/libutils.a` and write that dependency in the top level makefile so by the time the sub-make is trying to build an executable, the `libutils.a` has already been built. It'll work normally when invoked from the sub-directory.
2. get rid of the sub-directory makefiles (or write them in a way to be included from top-level) and never use recursive make. This is less code in total and will perform a bit better, but doesn't readily support running `make` from the subdirectory (without some work). I don't really use that development mode myself, and maybe it's no great loss.

Comment (jeremylt) [reply examples/petsc/Makefile:48]: Ultimately, I'd like to make 1) work across all of the petsc based examples.

Comment (jedbrown) [reply examples/petsc/Makefile]: I still don't love the recursive make, but it should be reliable after my push.

**Comment by jeremylt** [comment]
This leaves 
- [ ] lcov
- [ ] Possibly adding solids + reduced fluids tests
as the open ToDos?

**Comment by jedbrown** [comment]
Yeah, need to install lcov from source. Let's merge this and do CUDA tests for fluids and solids in a new PR after #754 merges.

**Review by jeremylt — APPROVED** [review]
